controllers/suite_test.go

```diff
@@ -17,95 +17,63 @@ limitations under the License.
-	// +kubebuilder:scaffold:imports
 )
 
-// These tests use Ginkgo (BDD-style Go testing framework). Refer to
-// http://onsi.github.io/ginkgo/ to learn more about Ginkgo.
-
-var cfg *rest.Config
-var k8sClient client.Client
-var k8sManager ctrl.Manager
-var imageAutoReconciler *ImageUpdateAutomationReconciler
-var testEnv *envtest.Environment
-
-func TestAPIs(t *testing.T) {
-	RegisterFailHandler(Fail)
-
-	RunSpecsWithDefaultAndCustomReporters(t,
-		"Controller Suite",
-		[]Reporter{printer.NewlineReporter{}})
-}
-
-var _ = BeforeSuite(func(done Done) {
-	ctrl.SetLogger(
-		zap.New(zap.WriteTo(GinkgoWriter), zap.UseDevMode(true)),
-	)
-
-	By("bootstrapping test environment")
-	testEnv = &envtest.Environment{
-		CRDDirectoryPaths: []string{
-			filepath.Join("..", "config", "crd", "bases"),
-			filepath.Join("testdata", "crds"),
-		},
-	}
-
-	var err error
-	cfg, err = testEnv.Start()
-	Expect(err).ToNot(HaveOccurred())
-	Expect(cfg).ToNot(BeNil())
+const (
+	contextTimeout = time.Second * 10
+)
 
-	Expect(sourcev1.AddToScheme(scheme.Scheme)).To(Succeed())
-	Expect(imagev1_reflect.AddToScheme(scheme.Scheme)).To(Succeed())
+var (
+	testEnv *testenv.Environment
+	ctx     = ctrl.SetupSignalHandler()
+)
 
-	Expect(imagev1.AddToScheme(scheme.Scheme)).To(Succeed())
-	// +kubebuilder:scaffold:scheme
+func TestMain(m *testing.M) {
+	utilruntime.Must(imagev1_reflect.AddToScheme(scheme.Scheme))
+	utilruntime.Must(sourcev1.AddToScheme(scheme.Scheme))
+	utilruntime.Must(imagev1.AddToScheme(scheme.Scheme))
 
-	k8sManager, err = ctrl.NewManager(cfg, ctrl.Options{
-		Scheme: scheme.Scheme,
-	})
-	Expect(err).ToNot(HaveOccurred())
+	testEnv = testenv.New(testenv.WithCRDPath(
+		filepath.Join("..", "config", "crd", "bases"),
+		filepath.Join("testdata", "crds"),
+	))
 
-	imageAutoReconciler = &ImageUpdateAutomationReconciler{
-		Client: k8sManager.GetClient(),
+	if err := (&ImageUpdateAutomationReconciler{
+		Client: testEnv,
 		Scheme: scheme.Scheme,
+	}).SetupWithManager(testEnv, ImageUpdateAutomationReconcilerOptions{}); err != nil {
+		panic(fmt.Sprintf("Failed to start ImageUpdateAutomationReconciler: %v", err))
 	}
-	Expect(imageAutoReconciler.SetupWithManager(k8sManager, ImageUpdateAutomationReconcilerOptions{})).To(Succeed())
 
 	go func() {
-		defer GinkgoRecover()
-		err = k8sManager.Start(ctrl.SetupSignalHandler())
-		Expect(err).ToNot(HaveOccurred())
+		fmt.Println("Starting the test environment")
+		if err := testEnv.Start(ctx); err != nil {
+			panic(fmt.Sprintf("Failed to start the test environment manager: %v", err))
+		}
 	}()
+	<-testEnv.Manager.Elected()
 
-	// Specifically an uncached client. Use <reconciler>.Get if you
-	// want to see what the reconcilers see.
-	k8sClient, err = client.New(cfg, client.Options{Scheme: scheme.Scheme})
-	Expect(err).ToNot(HaveOccurred())
-	Expect(k8sClient).ToNot(BeNil())
+	code := m.Run()
 
-	close(done)
-}, 60)
+	fmt.Println("Stopping the test environment")
+	if err := testEnv.Stop(); err != nil {
```

Comment:
This hangs when used with controller-runtime > 0.10, we need to stop the controller before testenv. Here is an example: https://github.com/fluxcd/kustomize-controller/blob/v1beta2/controllers/suite_test.go#L150-L151

Reply:
Hi, I couldn't reproduce that. I used to see similar issues before, when using envtest. But since the refactor, I haven't seen that.
Also, I tried updating to controller-runtime separately, just to reproduce it and instead of the hang issue, I discovered some other failures due to a breaking change in the controller-runtime client, refer: https://github.com/kubernetes-sigs/controller-runtime/pull/1640.
I fixed that and everything works well with CR v0.10.1. I used to be able to run `pgrep` to look for `etcd` and `apiserver` when it used to experience the hanging behavior. I don't get any results anymore. Seems to be stopping everything correctly.
Could it be something to do with the environment? I tried this on a linux environment.

Reply:
Update: I create #230 to update controller-runtime separately. While running the tests in that branch, I was able to reproduce the hanging behavior. Although the tests passed, the overall tests failed because the kube-apiserver couldn't be stopped:

```console
  Unexpected error:
      <errors.aggregate | len:1, cap:1>: [
          <*errors.errorString | 0xc0005085d0>{
              s: "timeout waiting for process kube-apiserver to stop",
          },
      ]
      timeout waiting for process kube-apiserver to stop
  occurred
```

Again, I can't reproduce it with these refactored tests.

Reply:
Even if it no longer hangs, I think we should close the controller before envtest, otherwise the controller will write dozen error logs to stderr due to failing to connect to the API.